beacon_node/operation_pool/src/attester_slashing.rs

```diff
@@ -0,0 +1,71 @@
+use crate::max_cover::MaxCover;
+use std::collections::HashMap;
+use types::{AttesterSlashing, BeaconState, ChainSpec, EthSpec};
+
+pub struct AttesterSlashingMaxCover<'a, T: EthSpec> {
+    slashing: &'a AttesterSlashing<T>,
+    effective_balances: HashMap<u64, u64>,
+}
+
+impl<'a, T: EthSpec> AttesterSlashingMaxCover<'a, T> {
+    pub fn new(
+        slashing: &'a AttesterSlashing<T>,
+        state: &BeaconState<T>,
+        spec: &ChainSpec,
+    ) -> Option<Self> {
+        let length = slashing.attestation_1.attesting_indices.len()
+            + slashing.attestation_2.attesting_indices.len();
+        let mut effective_balances: HashMap<u64, u64> = HashMap::with_capacity(length);
+
+        for vd in &slashing.attestation_1.attesting_indices {
+            let eff_balance = state.get_effective_balance(*vd as usize, spec).ok()?;
+            effective_balances.insert(*vd, eff_balance);
+        }
+
+        for vd in &slashing.attestation_2.attesting_indices {
+            let eff_balance = state.get_effective_balance(*vd as usize, spec).ok()?;
+            effective_balances.insert(*vd, eff_balance);
+        }
```

Comment:
At the moment this adds all the validators from the union of the two indexed attestations to the map, when it should add only the slashable validators.

I think we should move the call to `get_slashable_indices_modular` from `lib.rs` into this function, and then only add those validators to the map. This will require the list of validators that have already been slashed by the proposer slashings (currently called `to_be_slashed`), but just as an immutable reference, because the max cover algorithm takes care of the rest of the changes.

(sorry pressed submit too early before and lost some work, let me know if you need more clarity on this)

Reply:
I think i understand? you wish to transfer the whole filter() block logic into the AttesterSlashingMaxCover::new() which is fine, I can do that.  Just curious what you're thinking, why?

Reply:
> you wish to transfer the whole filter() block logic into the AttesterSlashingMaxCover::new() which is fine, I can do that

Pretty much, yeah, but without the part where we mutate the `to_be_slashed` set, i.e. don't bring this block along:

https://github.com/sigp/lighthouse/blob/b185d7bbd84deade495dbb5de4a57ff42d544194/beacon_node/operation_pool/src/lib.rs#L238-L245

That logic is a poor attempt at doing what max coverage will do a lot better, and we don't want it interfering.

However, the main issue with the two loops above is that if we have a slashing with indices like this:

```
attestation_1: {1, 2, 3, 4, 5, 6},
attestation_2: {6, 7, 8, 9, 10}
```

then it will add the effective balances of all validators 1-10 to the map, instead of just the effective balance of validator 6 (because 6 is the only validator slashed by that slashing). The `get_slashing_indices_modular` function (from the filter), is exactly the function for computing this set intersection.